ledger/acctupdates.go

```diff
@@ -288,12 +325,253 @@ func (au *accountUpdates) close() {
 	}
 	au.baseAccounts.prune(0)
 	au.baseResources.prune(0)
+	au.baseKVs.prune(0)
 }
 
 func (au *accountUpdates) LookupResource(rnd basics.Round, addr basics.Address, aidx basics.CreatableIndex, ctype basics.CreatableType) (ledgercore.AccountResource, basics.Round, error) {
 	return au.lookupResource(rnd, addr, aidx, ctype, true /* take lock */)
 }
 
+func (au *accountUpdates) LookupKv(rnd basics.Round, key string) (*string, error) {
+	return au.lookupKv(rnd, key, true /* take lock */)
+}
+
+func (au *accountUpdates) lookupKv(rnd basics.Round, key string, synchronized bool) (*string, error) {
+	needUnlock := false
+	if synchronized {
+		au.accountsMu.RLock()
+		needUnlock = true
+	}
+	defer func() {
+		if needUnlock {
+			au.accountsMu.RUnlock()
+		}
+	}()
+
+	// TODO: This loop and round handling is copied from other routines like
+	// lookupResource. I believe that it is overly cautious, as it always reruns
+	// the lookup if the DB round does not match the expected round. However, as
+	// long as the db round has not advanced too far (greater than `rnd`), I
+	// believe it would be valid to use. In the interest of minimizing changes,
+	// I'm not doing that now.
+
+	for {
+		currentDbRound := au.cachedDBRound
+		currentDeltaLen := len(au.deltas)
+		offset, err := au.roundOffset(rnd)
+		if err != nil {
+			return nil, err
+		}
+
+		// check if we have this key in `kvStore`, as that means the change we
+		// care about is in kvDeltas (and maybe just kvStore itself)
+		mval, indeltas := au.kvStore[key]
+		if indeltas {
+			// Check if this is the most recent round, in which case, we can
+			// use a cache of the most recent kvStore state
+			if offset == uint64(len(au.kvDeltas)) {
+				return mval.data, nil
+			}
+
+			// the key is in the deltas, but we don't know if it appears in the
+			// delta range of [0..offset-1], so we'll need to check. Walk deltas
+			// backwards so later updates take priority.
+			for i := int(offset - 1); i >= 0; i-- {
```

Comment:
nit: I'd prefer the same loop style like existing lookup functions have
```
for offset > 0 {
    offset--
    ...
}

Reply:
Added into the byteslice conversion.

https://github.com/algorand/go-algorand/pull/4686